First Review

bradt opened the pull request.

bradt committed:
Adding for review

bradt commented:
@lewis-dbi Oh man, this really looks great. The thing I love the most is that I didn't find one thing out of place when looking at it in the browser. In my experience, that's been most of the battle I've had reviewing frontend work. Devs just don't see the inconsistencies between the mockups and what their HTML+CSS is producing, so it's really great not to have screenshot things and point out the subtle differences.

I've created this Pull Request so that I can add comments to specific lines of code and we can discuss them and loop Pete in as well. I'll submit a review soon.

1ewiswarren committed:
Update components.html

1ewiswarren committed:
Update style.css

bradt commented on css/style.css:
```diff
@@ -0,0 +1,1215 @@
+  font-weight: 300;
+  font-size: 1.5rem;
+  margin-bottom: 2rem;
+}
```
Ideally we would be able to just include this stylesheet as is as part of the product and it would just work, but the styles above are just for displaying the style guide. I would put them in a separate stylesheet.

bradt commented on components.html:
```diff
@@ -0,0 +1,1473 @@
+
+
+    <!--COMPONENT MARGIN AND LABEL-->
+    <div class="component-margin"></div>
```
Rather than have an HTML element here just for spacing, I would have just added margin/padding to the `.component-label` class.

bradt commented on components.html:
```diff
@@ -0,0 +1,1473 @@
+    <!--NAV-->
+    <ul class="nav bg-white flex-container">
+      <li class="nav-item nav-item-active">
+        <a>Profiles</a>
```
I would add `href=""` to all the links because that attribute changes the behaviour of the `<a>` tag. I believe it adds a dotted outline when clicked and you might want to hide that.

bradt commented on components.html:
```diff
@@ -0,0 +1,1473 @@
+    <!--ACCORDION COLLAPSED-->
+
+    <div class="accordion-collapsed container-shadow bg-white flex-container">
+      <div style="width:10rem;" class="flex-container">
```
Why the inline style here?

bradt commented on components.html:
```diff
@@ -0,0 +1,1473 @@
+    <!--COMPONENT MARGIN AND LABEL-->
+    <div class="component-margin"></div>
+    <p class="component-label">Accordion States</p>
+    <!--COMPONENT MARGIN-->
```
I think these accordions require a min-width setting. When I make the screen narrow they become horribly distorted...

![image](https://user-images.githubusercontent.com/190095/62389866-511a0800-b537-11e9-88e3-529adba2257b.png)

It would be better if they stopped shrinking at a certain point and then you have to horizontally scroll.

bradt commented on components.html:
```diff
@@ -0,0 +1,1473 @@
+
+    <!--Default Action-->
+    <div class="flex-container">
+      <div class="action-icon flex-container">
```
Maybe this should be a `<button>` tag?

bradt commented on components.html:
```diff
@@ -0,0 +1,1473 @@
+          <path fill="#04223F"
+            d="M18.1497393 17.0344828C20.2470977 17.0344828 21.9443022 16.5012864 21.954651 15.846421L23.3103448 15.8428425C23.3103448 15.1915557 22.7032148 14.6655163 21.954651 14.6655163 21.2095368 14.6655163 20.6058564 15.1915557 20.6024068 15.839264L19.6917117 15.839264 22.0477903 11.1371162C20.5127171 10.3104829 18.6499317 10.8830427 17.8806703 12.4146404 17.6909421 12.7975398 17.5840045 13. ...
+        </g>
+      </svg>
```
Embedding an SVG is ok if it only ever appears once in one place on one page. For example, embedding a diagram on a sales page that's only ever show there is fine. But for SVGs that appear across page loads and in multiple places, it probably worth combining them all into a sprite map.

bradt commented on components.html:
```diff
@@ -0,0 +1,1473 @@
+        <p class="container-table-text">Content goes here</p>
+        <p class="container-table-text">Content goes here</p>
+      </tr>
+    </div>
```
I don't think this is valid HTML because of the `<tr>`. I would probably just use a `<ul>` for this component.

bradt commented on components.html:
```diff
@@ -0,0 +1,1473 @@
+      <div class="flex-container">
+
+        <h1 class="migration-title">
+          Push <span class="regular">to </span>test1.lewisbrains.us</span>
```
```suggestion
          Push <span class="regular">to </span>test1.lewisbrains.us
```

bradt commented on components.html:
```diff
@@ -0,0 +1,1473 @@
+              <div class="align-right">
+                <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" viewBox="0 0 16 16">
+                  <path fill="#A5DDF1"
+                    d="M8.57952685,1.37620917 L10.2429656,4.90369756 C10.3369312,5.10295441 10.5185555,5.24105328 10.7286459,5.27298678 L14.4476796,5.83873515 C14.6907017,5.87567445 14.8925991,6.05385549 14.968446,6.29832852 C15.0442929,6.54280156 14.9809285,6.8111458 14.8050061,6.99048543 L12.1134474,9.73618788 C11.9612289,9.89146488 11.8922145,10.1149287 11.9276892,10.3336668 L12.563008,14.2108 ...
```
```suggestion
          <td class="table-col-date">
```

If you have to update the flex value here, you have to update it for every row. Probably better to give a class to this that can be updated in one place. The additional advantage is that you can then apply further styles to cells in that column.

bradt replied:
Rather than have an empty HTML element here just for the divider, I would typically wrap the entire bottom part in a div and then just give that a top border.

bradt reviewed (commented):
Markup looks very good. Just a few pointers. I'll review the CSS next.

Commit:
Create .gitignore

Added gitignore to exclude node_modules and compiled assets

Commit:
Added Gulp task runner

For SVG minifying and sprite sheet

Commit:
Created Layout stylesheet

This is for layout only styling to display the components on the page including labels and margins.

Commit:
Update style.css

Removed layout only styling and amended CSS as per feedback

Commit:
Update components.html

Added inline SVGs using a spritesheet, added buttons tags to all components that required them and other feedback applied.

Commit:
Removed unused dependancies

Removed unused dependencies

1ewiswarren commented:
@bradt Ready to review. I made quite a few changes.

- Fixed all amends above.
- Using SVG spritesheet instead of embedding them in the page
- Found some other elements that required <button tags>

Commit:
Update components.html

Added a .wpmdb class so that the styling doesnt get overwritten in the WP dashboard

Commit:
Update components.html

Fixed a few errors and adjust some items for the WP dashboard.

Commit:
Update component-layout.css

Stirctly for laying out the components on the page. Wrapped inside a .wpmdb tag

Commit:
Removed

Commit:
Updated CSS. Using Sass

Sass added to this project and this is the CSS file wrapped inside .wpmdb tag

Commit:
WP overrides css file

This styling overrides some other styling when placed inside the WP dashboard.

Commit:
Added Sass

Commit:
Create favicon.png

Commit:
Added Sass

Commit:
Sass files for styling

1ewiswarren commented:
@bradt All updated and as close as I can get it inside the WP dashboard. I will send you a zip file separately on Monday so you can how it looks inside the WP dashboard.

One thing I'm still not sure about is if we should set a min-width to a whole wrapper tag that sits around the whole codebase or just on individual components like the accordions and the nav etc. 

Also, using SASS now too. Although it's basically my CSS nested inside a .wpmdb tag.

bradt commented on scss/wp-overrides.scss:
```diff
@@ -0,0 +1,77 @@
+
+.migration-complete-summary {
+  font-size: 1.06rem!important;
+}
```
IMO making these styles separate, in their own file makes things harder. Imagine you're looking at the `.migration-complete-summary` class in style.scss and wracking your brain trying to figure out why the font size is showing as `!important`, forgetting about this overrides file. Why not put the `!important` statement in the style.scss file?

1ewiswarren replied:
No worries. I thought it might be simpler to separate the files but maybe it isnt then. Ill add these back into style.scss.

1ewiswarren replied:
Should I append them to the end of style.scss or just place them where they would appear in the file?

bradt replied:
Place them where they should appear

bradt commented on scss/style.scss:
```diff
@@ -0,0 +1,1241 @@
+body {
+  font-family: system-ui, -apple-system, BlinkMacSystemFont, "Segoe UI", "Roboto", "Oxygen", "Ubuntu", "Cantarell", "Fira Sans", "Droid Sans", "Helvetica Neue", sans-serif;
+}
```
This should be part of `component-layout.scss` as we definitely wouldn't want to include this in the WP dashboard.

1ewiswarren replied:
@bradt Yeah good point. I think WP will use this anyway as it's default for system ui. Wil move to components-layout.scss

bradt replied:
Yes, but if WP updates its style our sheet could override it.

bradt commented on scss/style.scss:
```diff
@@ -0,0 +1,1241 @@
+  .wrapper {
+    max-width: 1280px;
+    margin: 100px;
+  }
```
This seems to be just for displaying the component library and not for use in the plugin styles? Should probably be part of `component-layout.scss`.

1ewiswarren replied:
How do you mean sorry? Are you saying this won't be used and is just for display purposes? So should be moved to component-layout.scss?

bradt replied:
Yes

bradt commented on scss/style.scss:
```diff
@@ -0,0 +1,1241 @@
+
+  .html {
+    font-size: 10px;
+  }
```
I can't find this class in the markup

1ewiswarren replied:
This is actually a mistake. Originally I was trying to set the html (not .html) to a font-size of 10px as a baseline for my rem sizing.  Ill remove.

bradt commented:
This looks great Lewis, nice work! 🎉 Only a few minor things for you to review here. I expect that Pete will have some more pointed feedback for you once he starts implementing things.

bradt commented:
Once you've finished with this round of feedback, please give Pete the nod to start implementing.

Commit:
Removed file

Added contents to styles.scss

Commit:
Added layout only styling

Commit:
Removed file and added contents to styles.scss

Commit:
Revert "Removed file and added contents to styles.scss"

This reverts commit feaedc3547d7119eaf34deafdc33e6de1f105957.

Commit:
Added gulp watch task

Commit:
Updated styling based on feedback

1ewiswarren commented:
@ptasker Ready for you to use.

Commit:
Updated description

Just testing out where my avatar now shows when doing a commit

ptasker commented on gulpfile.js:
```diff
@@ -0,0 +1,38 @@
+var gulp = require('gulp');
```
Looks like you need to add a 'default' task for Gulp here. Probably just run the `svgstore` and `sass` tasks.

ptasker replied:
https://subscription.packtpub.com/book/application_development/9781787283732/4/ch04lvl2sec28/the-default-task

ptasker commented on package.json:
```diff
@@ -0,0 +1,19 @@
+    "test": "echo \"Error: no test specified\" && exit 1"
+  },
+  "author": "",
+  "license": "ISC",
```
Need to remove the comma here, causes `npm install` to fail

ptasker commented on scss/style.scss:
```diff
@@ -0,0 +1,1259 @@
+    justify-content: space-between;
+  }
+  /*Nested Expanded*/
+  .nested-accordion {
```
Between this class and the `.nested-accordion-expanded` below there are few rules duplicated. I would extend this class with something like

```css
.nested-accordion {
    ...
    &.expanded{
      background: #45719B;
    }
  }
```

So the accordion div would look like:
//Closed
`<div class="nested-accordion">`

//Expanded
`<div class="nested-accordion expanded">`

In any event, less duplicated styles means less CSS in general which is always a good thing.

bradt replied:
Agreed, less duplication is better. Only problem with your proposal is that if you define a general `expanded` class later, it's going to affect the styles here. In the past I've found it better to be more specific with class naming in cases like this, so keep the class name `nested-accordion-expanded` but only include the additional styles for the expanded state and do

    <div class="nested-accordion nested-accordion-expanded">

ptasker commented on components.html:
```diff
@@ -0,0 +1,1381 @@
+
+      <!--NAV-->
+      <ul class="nav bg-white flex-container">
+        <li class="nav-item nav-item-active">
```
I don't think each `<li>` needs a classname applied. Can be targeted with `ul.nav li`.

ptasker commented on components.html:
```diff
@@ -0,0 +1,1381 @@
+      <!--SUB NAV-->
+
+      <ul class="subnav flex-container">
+        <li class="subnav-item subnav-item-active">
```
Same note as above, no need for the '.subnav' item class here

ptasker commented on components.html:
```diff
@@ -0,0 +1,1381 @@
+
+      <!--Toggle-->
+      <div class="switch_box box_1">
+        <input type="checkbox" class="switch_1">
```
using both underscores and dashes for classnames, should probably standardize on one or the other.

ptasker commented on components.html:
```diff
@@ -0,0 +1,1381 @@
+
+
+      <!--Speech bubble (button tooltip)-->
+      <div class="flex-container">
```
Different HTML elements could help here to reduce the 'div-itis' 😂

1ewiswarren replied:
@bradt Is there a general practice for this? Like particular elements to use?

bradt replied:
Personally, I don't think it matters at all for an app like this that is never going to be crawled by a search engine. If you look at the markup behind Gmail, it's all divs. If you're doing markup for our sites though, you definitely want to be using the right HTML5 elements for the right situations as Google's crawler rewards that. [Mozilla's HTML5 guide](https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/HTML5) seems pretty good at a glance, but there might be a better one out there.

ptasker commented on components.html:
```diff
@@ -0,0 +1,1381 @@
+            <input type="checkbox" class="checkbox-default" name="" value="">
+            <label>File path goes here</label>
+          </td>
+          <td style="flex-grow:2;">
```
Is there a reason for the inline styles here?

1ewiswarren replied:
I did some inline styles (flex-grow properties) in various places as there are quite a few columns in some of the layouts and I didn't know if it was worth settings classes for so many columns, particularly when some columns are just spacings.

@bradt what do you think?

bradt replied:
I would say that inline styles make sense only when the JS is going to be altering it dynamically. For example, in a progress bar, the JS might manipulate the width. Otherwise it should be a class so that if additional styles need to be applied to that element in the future it can be done relatively easily by editing the stylesheet.

1ewiswarren replied:
All fixed up now and committed.

ptasker commented on components.html:
```diff
@@ -0,0 +1,1381 @@
+
+
+      <!--Search and Replace-->
+      <table>
```
In my current markup I'm avoiding using the table element as it's not really a dataset, which is what `<table>` is typically used for these days.

1ewiswarren replied:
@bradt Any ideas for this? I replaced the <table> tag with a div and it breaks the layout. I tried a few different things but couldn't get it to work.

bradt replied:
Again, I don't think semantics matter much here, so it's really whatever HTML and CSS are the best fit for the job. @ptasker What's your suggestion?

ptasker replied:
In the placeholder markup I'm using CSS Grid with React's styled components  with a wrapper div as below:

```javascript
export const Row  = styled.div`
   display: grid;
   margin: 0.4rem 0;
   grid-template-columns: 0.5fr 5fr 0.5fr 5fr 0.5fr;
   row-gap: 10px;
   column-gap: 10px;
      .header-2{
          grid-column: 2
      }
      
      .header-4{
          grid-column: 4
      }
`;
```

This guy wraps 5 divs, and it creates the same layout as a table. Much less markup.
https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Grid_Layout

I can make the styles you have here work with the existing markup I have, so no worries updating this.

1ewiswarren replied:
@ptasker Yes please. Then I'll have a look at it so I know for next time.

bradt replied:
@lewis-dbi Maybe you ought to convert the component library from Flexbox to CSS Grid? Sounds like something for you that's worth learning.

ptasker commented on components.html:
```diff
@@ -0,0 +1,1381 @@
+      <!--Table container (multi-select list)-->
+      <div class="container-table">
+        <ul class="container-table-row">
+          <li class="container-table-text">
```
No need for classname, can just be `.container-table-row li`

ptasker reviewed (commented):
Hey @lewis-dbi, looking good. I just noticed a few things when going to start implementing.

Commit:
Fixed comma issue

Commit:
Added a default task

Commit:
Removed overrides file

This rules are now in the main style.scss file

Commit:
Various amends to CSS

Removed duplicate css and made some other amends as per Pete's feedback

Commit:
HTML amends

Amended some of the HTML as per Pete's feedback. This includes remove some class names and targeting that element instead.

Commit:
Removed inline styles for all tables

Created a class inside of the sass file, and then nested .td classes insidethe corresponding class.

decided not to name classes as they are pretty arbitrary.
